File: python/mirascope/ops/_internal/instrumentation/llm/response.py
"""Mirascope-specific instrumentation for Response.resume methods."""

from __future__ import annotations

from functools import wraps
from typing import TYPE_CHECKING, Any, cast, overload

from .....llm import (
    AsyncContextResponse,
    AsyncContextStreamResponse,
    AsyncResponse,
    AsyncStreamResponse,
    Context,
    ContextResponse,
    ContextStreamResponse,
    DepsT,
    FormattableT,
    Response,
    StreamResponse,
    UserContent,
)
from ...spans import Span as MirascopeSpan
from .serialize import attach_mirascope_response, attach_mirascope_response_async

if TYPE_CHECKING:
    pass


# =============================================================================
# Original method references and wrapped state flags
# =============================================================================

_ORIGINAL_RESPONSE_RESUME = Response.resume
_RESPONSE_RESUME_WRAPPED = False
_ORIGINAL_ASYNC_RESPONSE_RESUME = AsyncResponse.resume
_ASYNC_RESPONSE_RESUME_WRAPPED = False
_ORIGINAL_CONTEXT_RESPONSE_RESUME = ContextResponse.resume
_CONTEXT_RESPONSE_RESUME_WRAPPED = False
_ORIGINAL_ASYNC_CONTEXT_RESPONSE_RESUME = AsyncContextResponse.resume
_ASYNC_CONTEXT_RESPONSE_RESUME_WRAPPED = False
_ORIGINAL_STREAM_RESPONSE_RESUME = StreamResponse.resume
_STREAM_RESPONSE_RESUME_WRAPPED = False
_ORIGINAL_ASYNC_STREAM_RESPONSE_RESUME = AsyncStreamResponse.resume
_ASYNC_STREAM_RESPONSE_RESUME_WRAPPED = False
_ORIGINAL_CONTEXT_STREAM_RESPONSE_RESUME = ContextStreamResponse.resume
_CONTEXT_STREAM_RESPONSE_RESUME_WRAPPED = False
_ORIGINAL_ASYNC_CONTEXT_STREAM_RESPONSE_RESUME = AsyncContextStreamResponse.resume
_ASYNC_CONTEXT_STREAM_RESPONSE_RESUME_WRAPPED = False


# =============================================================================
# Response.resume instrumentation
# =============================================================================


@overload
def _instrumented_response_resume(self: Response, content: UserContent) -> Response: ...


@overload
def _instrumented_response_resume(
    self: Response[FormattableT], content: UserContent
) -> Response[FormattableT]: ...


@wraps(_ORIGINAL_RESPONSE_RESUME)
def _instrumented_response_resume(
    self: Response | Response[FormattableT], content: UserContent
) -> Response | Response[FormattableT]:
    """Returns a Mirascope-traced result of `Response.resume`."""
    with MirascopeSpan(f"Response.resume {self.model_id}") as span:
        span.set(
            **{
                "mirascope.type": "response_resume",
                "mirascope.response.model_id": self.model_id,
                "mirascope.response.provider_id": self.provider_id,
            }
        )
        result = cast(
            "Response | Response[FormattableT]",
            _ORIGINAL_RESPONSE_RESUME(cast(Any, self), content),
        )
        attach_mirascope_response(span, result)
        return result


def wrap_response_resume() -> None:
    """Returns None. Replaces `Response.resume` with the instrumented wrapper."""
    global _RESPONSE_RESUME_WRAPPED
    if _RESPONSE_RESUME_WRAPPED:
        return
    Response.resume = _instrumented_response_resume
    _RESPONSE_RESUME_WRAPPED = True


def unwrap_response_resume() -> None:
    """Returns None. Restores the original `Response.resume` implementation."""
    global _RESPONSE_RESUME_WRAPPED
    if not _RESPONSE_RESUME_WRAPPED:
        return
    Response.resume = _ORIGINAL_RESPONSE_RESUME
    _RESPONSE_RESUME_WRAPPED = False


# =============================================================================
# AsyncResponse.resume instrumentation
# =============================================================================


@overload
async def _instrumented_async_response_resume(
    self: AsyncResponse, content: UserContent
) -> AsyncResponse: ...


@overload
async def _instrumented_async_response_resume(
    self: AsyncResponse[FormattableT], content: UserContent
) -> AsyncResponse[FormattableT]: ...


@wraps(_ORIGINAL_ASYNC_RESPONSE_RESUME)
async def _instrumented_async_response_resume(
    self: AsyncResponse | AsyncResponse[FormattableT], content: UserContent
) -> AsyncResponse | AsyncResponse[FormattableT]:
    """Returns a Mirascope-traced result of `AsyncResponse.resume`."""
    with MirascopeSpan(f"AsyncResponse.resume {self.model_id}") as span:
        span.set(
            **{
                "mirascope.type": "response_resume",
                "mirascope.response.model_id": self.model_id,
                "mirascope.response.provider_id": self.provider_id,
            }
        )
        result = cast(
            "AsyncResponse | AsyncResponse[FormattableT]",
            await _ORIGINAL_ASYNC_RESPONSE_RESUME(cast(Any, self), content),
        )
        await attach_mirascope_response_async(span, result)
        return result


def wrap_async_response_resume() -> None:
    """Returns None. Replaces `AsyncResponse.resume` with the instrumented wrapper."""
    global _ASYNC_RESPONSE_RESUME_WRAPPED
    if _ASYNC_RESPONSE_RESUME_WRAPPED:
        return
    AsyncResponse.resume = _instrumented_async_response_resume
    _ASYNC_RESPONSE_RESUME_WRAPPED = True


def unwrap_async_response_resume() -> None:
    """Returns None. Restores the original `AsyncResponse.resume` implementation."""
    global _ASYNC_RESPONSE_RESUME_WRAPPED
    if not _ASYNC_RESPONSE_RESUME_WRAPPED:
        return
    AsyncResponse.resume = _ORIGINAL_ASYNC_RESPONSE_RESUME
    _ASYNC_RESPONSE_RESUME_WRAPPED = False


# =============================================================================
# ContextResponse.resume instrumentation
# =============================================================================


@overload
def _instrumented_context_response_resume(
    self: ContextResponse[DepsT], ctx: Context[DepsT], content: UserContent
) -> ContextResponse[DepsT]: ...


@overload
def _instrumented_context_response_resume(
    self: ContextResponse[DepsT, FormattableT],
    ctx: Context[DepsT],
    content: UserContent,
) -> ContextResponse[DepsT, FormattableT]: ...


@wraps(_ORIGINAL_CONTEXT_RESPONSE_RESUME)
def _instrumented_context_response_resume(
    self: ContextResponse[DepsT] | ContextResponse[DepsT, FormattableT],
    ctx: Context[DepsT],
    content: UserContent,
) -> ContextResponse[DepsT] | ContextResponse[DepsT, FormattableT]:
    """Returns a Mirascope-traced result of `ContextResponse.resume`."""
    with MirascopeSpan(f"ContextResponse.resume {self.model_id}") as span:
        span.set(
            **{
                "mirascope.type": "response_resume",
                "mirascope.response.model_id": self.model_id,
                "mirascope.response.provider_id": self.provider_id,
            }
        )
        result = cast(
            "ContextResponse[DepsT] | ContextResponse[DepsT, FormattableT]",
            _ORIGINAL_CONTEXT_RESPONSE_RESUME(cast(Any, self), ctx, content),
        )
        attach_mirascope_response(span, result)
        return result


def wrap_context_response_resume() -> None:
    """Returns None. Replaces `ContextResponse.resume` with the instrumented wrapper."""
    global _CONTEXT_RESPONSE_RESUME_WRAPPED
    if _CONTEXT_RESPONSE_RESUME_WRAPPED:
        return
    ContextResponse.resume = _instrumented_context_response_resume
    _CONTEXT_RESPONSE_RESUME_WRAPPED = True


def unwrap_context_response_resume() -> None:
    """Returns None. Restores the original `ContextResponse.resume` implementation."""
    global _CONTEXT_RESPONSE_RESUME_WRAPPED
    if not _CONTEXT_RESPONSE_RESUME_WRAPPED:
        return
    ContextResponse.resume = _ORIGINAL_CONTEXT_RESPONSE_RESUME
    _CONTEXT_RESPONSE_RESUME_WRAPPED = False


# =============================================================================
# AsyncContextResponse.resume instrumentation
# =============================================================================


@overload
async def _instrumented_async_context_response_resume(
    self: AsyncContextResponse[DepsT], ctx: Context[DepsT], content: UserContent
) -> AsyncContextResponse[DepsT]: ...


@overload
async def _instrumented_async_context_response_resume(
    self: AsyncContextResponse[DepsT, FormattableT],
    ctx: Context[DepsT],
    content: UserContent,
) -> AsyncContextResponse[DepsT, FormattableT]: ...


@wraps(_ORIGINAL_ASYNC_CONTEXT_RESPONSE_RESUME)
async def _instrumented_async_context_response_resume(
    self: AsyncContextResponse[DepsT] | AsyncContextResponse[DepsT, FormattableT],
    ctx: Context[DepsT],
    content: UserContent,
) -> AsyncContextResponse[DepsT] | AsyncContextResponse[DepsT, FormattableT]:
    """Returns a Mirascope-traced result of `AsyncContextResponse.resume`."""
    with MirascopeSpan(f"AsyncContextResponse.resume {self.model_id}") as span:
        span.set(
            **{
                "mirascope.type": "response_resume",
                "mirascope.response.model_id": self.model_id,
                "mirascope.response.provider_id": self.provider_id,
            }
        )
        result = cast(
            "AsyncContextResponse[DepsT] | AsyncContextResponse[DepsT, FormattableT]",
            await _ORIGINAL_ASYNC_CONTEXT_RESPONSE_RESUME(
                cast(Any, self), ctx, content
            ),
        )
        await attach_mirascope_response_async(span, result)
        return result


def wrap_async_context_response_resume() -> None:
    """Returns None. Replaces `AsyncContextResponse.resume` with the instrumented wrapper."""
    global _ASYNC_CONTEXT_RESPONSE_RESUME_WRAPPED
    if _ASYNC_CONTEXT_RESPONSE_RESUME_WRAPPED:
        return
    AsyncContextResponse.resume = _instrumented_async_context_response_resume
    _ASYNC_CONTEXT_RESPONSE_RESUME_WRAPPED = True


def unwrap_async_context_response_resume() -> None:
    """Returns None. Restores the original `AsyncContextResponse.resume` implementation."""
    global _ASYNC_CONTEXT_RESPONSE_RESUME_WRAPPED
    if not _ASYNC_CONTEXT_RESPONSE_RESUME_WRAPPED:
        return
    AsyncContextResponse.resume = _ORIGINAL_ASYNC_CONTEXT_RESPONSE_RESUME
    _ASYNC_CONTEXT_RESPONSE_RESUME_WRAPPED = False


# =============================================================================
# StreamResponse.resume instrumentation
# =============================================================================


@overload
def _instrumented_stream_response_resume(
    self: StreamResponse, content: UserContent
) -> StreamResponse: ...


@overload
def _instrumented_stream_response_resume(
    self: StreamResponse[FormattableT], content: UserContent
) -> StreamResponse[FormattableT]: ...


@wraps(_ORIGINAL_STREAM_RESPONSE_RESUME)
def _instrumented_stream_response_resume(
    self: StreamResponse | StreamResponse[FormattableT], content: UserContent
) -> StreamResponse | StreamResponse[FormattableT]:
    """Returns a Mirascope-traced result of `StreamResponse.resume`."""
    with MirascopeSpan(f"StreamResponse.resume {self.model_id}") as span:
        span.set(
            **{
                "mirascope.type": "response_resume",
                "mirascope.response.model_id": self.model_id,
                "mirascope.response.provider_id": self.provider_id,
            }
        )
        result = cast(
            "StreamResponse | StreamResponse[FormattableT]",
            _ORIGINAL_STREAM_RESPONSE_RESUME(cast(Any, self), content),
        )
        attach_mirascope_response(span, result)
        return result


def wrap_stream_response_resume() -> None:
    """Returns None. Replaces `StreamResponse.resume` with the instrumented wrapper."""
    global _STREAM_RESPONSE_RESUME_WRAPPED
    if _STREAM_RESPONSE_RESUME_WRAPPED:
        return
    StreamResponse.resume = _instrumented_stream_response_resume
    _STREAM_RESPONSE_RESUME_WRAPPED = True


def unwrap_stream_response_resume() -> None:
    """Returns None. Restores the original `StreamResponse.resume` implementation."""
    global _STREAM_RESPONSE_RESUME_WRAPPED
    if not _STREAM_RESPONSE_RESUME_WRAPPED:
        return
    StreamResponse.resume = _ORIGINAL_STREAM_RESPONSE_RESUME
    _STREAM_RESPONSE_RESUME_WRAPPED = False


# =============================================================================
# AsyncStreamResponse.resume instrumentation
# =============================================================================


@overload
async def _instrumented_async_stream_response_resume(
    self: AsyncStreamResponse, content: UserContent
) -> AsyncStreamResponse: ...


@overload
async def _instrumented_async_stream_response_resume(
    self: AsyncStreamResponse[FormattableT], content: UserContent
) -> AsyncStreamResponse[FormattableT]: ...


@wraps(_ORIGINAL_ASYNC_STREAM_RESPONSE_RESUME)
async def _instrumented_async_stream_response_resume(
    self: AsyncStreamResponse | AsyncStreamResponse[FormattableT], content: UserContent
) -> AsyncStreamResponse | AsyncStreamResponse[FormattableT]:
    """Returns a Mirascope-traced result of `AsyncStreamResponse.resume`."""
    with MirascopeSpan(f"AsyncStreamResponse.resume {self.model_id}") as span:
        span.set(
            **{
                "mirascope.type": "response_resume",
                "mirascope.response.model_id": self.model_id,
                "mirascope.response.provider_id": self.provider_id,
            }
        )
        result = cast(
            "AsyncStreamResponse | AsyncStreamResponse[FormattableT]",
            await _ORIGINAL_ASYNC_STREAM_RESPONSE_RESUME(cast(Any, self), content),
        )
        await attach_mirascope_response_async(span, result)
        return result


def wrap_async_stream_response_resume() -> None:
    """Returns None. Replaces `AsyncStreamResponse.resume` with the instrumented wrapper."""
    global _ASYNC_STREAM_RESPONSE_RESUME_WRAPPED
    if _ASYNC_STREAM_RESPONSE_RESUME_WRAPPED:
        return
    AsyncStreamResponse.resume = _instrumented_async_stream_response_resume
    _ASYNC_STREAM_RESPONSE_RESUME_WRAPPED = True


def unwrap_async_stream_response_resume() -> None:
    """Returns None. Restores the original `AsyncStreamResponse.resume` implementation."""
    global _ASYNC_STREAM_RESPONSE_RESUME_WRAPPED
    if not _ASYNC_STREAM_RESPONSE_RESUME_WRAPPED:
        return
    AsyncStreamResponse.resume = _ORIGINAL_ASYNC_STREAM_RESPONSE_RESUME
    _ASYNC_STREAM_RESPONSE_RESUME_WRAPPED = False


# =============================================================================
# ContextStreamResponse.resume instrumentation
# =============================================================================


@overload
def _instrumented_context_stream_response_resume(
    self: ContextStreamResponse[DepsT], ctx: Context[DepsT], content: UserContent
) -> ContextStreamResponse[DepsT]: ...


@overload
def _instrumented_context_stream_response_resume(
    self: ContextStreamResponse[DepsT, FormattableT],
    ctx: Context[DepsT],
    content: UserContent,
) -> ContextStreamResponse[DepsT, FormattableT]: ...


@wraps(_ORIGINAL_CONTEXT_STREAM_RESPONSE_RESUME)
def _instrumented_context_stream_response_resume(
    self: ContextStreamResponse[DepsT] | ContextStreamResponse[DepsT, FormattableT],
    ctx: Context[DepsT],
    content: UserContent,
) -> ContextStreamResponse[DepsT] | ContextStreamResponse[DepsT, FormattableT]:
    """Returns a Mirascope-traced result of `ContextStreamResponse.resume`."""
    with MirascopeSpan(f"ContextStreamResponse.resume {self.model_id}") as span:
        span.set(
            **{
                "mirascope.type": "response_resume",
                "mirascope.response.model_id": self.model_id,
                "mirascope.response.provider_id": self.provider_id,
            }
        )
        result = cast(
            "ContextStreamResponse[DepsT] | ContextStreamResponse[DepsT, FormattableT]",
            _ORIGINAL_CONTEXT_STREAM_RESPONSE_RESUME(cast(Any, self), ctx, content),
        )
        attach_mirascope_response(span, result)
        return result


def wrap_context_stream_response_resume() -> None:
    """Returns None. Replaces `ContextStreamResponse.resume` with the instrumented wrapper."""
    global _CONTEXT_STREAM_RESPONSE_RESUME_WRAPPED
    if _CONTEXT_STREAM_RESPONSE_RESUME_WRAPPED:
        return
    ContextStreamResponse.resume = _instrumented_context_stream_response_resume
    _CONTEXT_STREAM_RESPONSE_RESUME_WRAPPED = True


def unwrap_context_stream_response_resume() -> None:
    """Returns None. Restores the original `ContextStreamResponse.resume` implementation."""
    global _CONTEXT_STREAM_RESPONSE_RESUME_WRAPPED
    if not _CONTEXT_STREAM_RESPONSE_RESUME_WRAPPED:
        return
    ContextStreamResponse.resume = _ORIGINAL_CONTEXT_STREAM_RESPONSE_RESUME
    _CONTEXT_STREAM_RESPONSE_RESUME_WRAPPED = False


# =============================================================================
# AsyncContextStreamResponse.resume instrumentation
# =============================================================================


@overload
async def _instrumented_async_context_stream_response_resume(
    self: AsyncContextStreamResponse[DepsT],
    ctx: Context[DepsT],
    content: UserContent,
) -> AsyncContextStreamResponse[DepsT]: ...


@overload
async def _instrumented_async_context_stream_response_resume(
    self: AsyncContextStreamResponse[DepsT, FormattableT],
    ctx: Context[DepsT],
    content: UserContent,
) -> AsyncContextStreamResponse[DepsT, FormattableT]: ...


@wraps(_ORIGINAL_ASYNC_CONTEXT_STREAM_RESPONSE_RESUME)
async def _instrumented_async_context_stream_response_resume(
    self: AsyncContextStreamResponse[DepsT]
    | AsyncContextStreamResponse[DepsT, FormattableT],
    ctx: Context[DepsT],
    content: UserContent,
) -> (
    AsyncContextStreamResponse[DepsT] | AsyncContextStreamResponse[DepsT, FormattableT]
):
    """Returns a Mirascope-traced result of `AsyncContextStreamResponse.resume`."""
    with MirascopeSpan(f"AsyncContextStreamResponse.resume {self.model_id}") as span:
        span.set(
            **{
                "mirascope.type": "response_resume",
                "mirascope.response.model_id": self.model_id,
                "mirascope.response.provider_id": self.provider_id,
            }
        )
        result = cast(
            "AsyncContextStreamResponse[DepsT] | AsyncContextStreamResponse[DepsT, FormattableT]",
            await _ORIGINAL_ASYNC_CONTEXT_STREAM_RESPONSE_RESUME(
                cast(Any, self), ctx, content
            ),
        )
        await attach_mirascope_response_async(span, result)
        return result


def wrap_async_context_stream_response_resume() -> None:
    """Returns None. Replaces `AsyncContextStreamResponse.resume` with the instrumented wrapper."""
    global _ASYNC_CONTEXT_STREAM_RESPONSE_RESUME_WRAPPED
    if _ASYNC_CONTEXT_STREAM_RESPONSE_RESUME_WRAPPED:
        return
    AsyncContextStreamResponse.resume = (
        _instrumented_async_context_stream_response_resume
    )
    _ASYNC_CONTEXT_STREAM_RESPONSE_RESUME_WRAPPED = True


def unwrap_async_context_stream_response_resume() -> None:
    """Returns None. Restores the original `AsyncContextStreamResponse.resume` implementation."""
    global _ASYNC_CONTEXT_STREAM_RESPONSE_RESUME_WRAPPED
    if not _ASYNC_CONTEXT_STREAM_RESPONSE_RESUME_WRAPPED:
        return
    AsyncContextStreamResponse.resume = _ORIGINAL_ASYNC_CONTEXT_STREAM_RESPONSE_RESUME
    _ASYNC_CONTEXT_STREAM_RESPONSE_RESUME_WRAPPED = False

File: typescript/src/ops/_internal/instrumentation/response.ts
/**
 * Mirascope-specific instrumentation for Response.resume methods.
 *
 * Provides wrapper functions to instrument Response resume calls with tracing,
 * matching the Python implementation pattern.
 */

import { context as otelContext, trace as otelTrace } from "@opentelemetry/api";

import type { Context } from "@/llm/context";
import type { UserContent } from "@/llm/messages";
import type { RootResponse } from "@/llm/responses/root-response";

import { ContextResponse } from "@/llm/responses/context-response";
import { ContextStreamResponse } from "@/llm/responses/context-stream-response";
import { Response } from "@/llm/responses/response";
import { StreamResponse } from "@/llm/responses/stream-response";
import { Span } from "@/ops/_internal/spans";
import { jsonStringify } from "@/ops/_internal/utils";

// =============================================================================
// Helper to attach response attributes
// =============================================================================

/**
 * Attach Mirascope-specific response attributes to a span.
 */
/* v8 ignore start - called from instrumented functions, tested in E2E */
function attachResponseAttributes(span: Span, response: RootResponse): void {
  span.set({
    "mirascope.trace.output": response.pretty(),
    "mirascope.response.messages": jsonStringify(
      response.messages
        .slice(0, -1)
        .map((m) => ({ role: m.role, content: m.content })),
    ),
    "mirascope.response.content": jsonStringify(response.content),
  });

  if (response.usage) {
    span.set({
      "mirascope.response.usage": jsonStringify({
        inputTokens: response.usage.inputTokens,
        outputTokens: response.usage.outputTokens,
        cacheReadTokens: response.usage.cacheReadTokens,
        cacheWriteTokens: response.usage.cacheWriteTokens,
        reasoningTokens: response.usage.reasoningTokens,
      }),
    });
  }
}
/* v8 ignore end */

// =============================================================================
// Original method references and wrapped state flags
// =============================================================================

// Response.resume
const _ORIGINAL_RESPONSE_RESUME = Response.prototype.resume;
let _RESPONSE_RESUME_WRAPPED = false;

// StreamResponse.resume
const _ORIGINAL_STREAM_RESPONSE_RESUME = StreamResponse.prototype.resume;
let _STREAM_RESPONSE_RESUME_WRAPPED = false;

// ContextResponse.resume
const _ORIGINAL_CONTEXT_RESPONSE_RESUME = ContextResponse.prototype.resume;
let _CONTEXT_RESPONSE_RESUME_WRAPPED = false;

// ContextStreamResponse.resume
const _ORIGINAL_CONTEXT_STREAM_RESPONSE_RESUME =
  ContextStreamResponse.prototype.resume;
let _CONTEXT_STREAM_RESPONSE_RESUME_WRAPPED = false;

// =============================================================================
// Response.resume instrumentation
// =============================================================================

/**
 * Instrumented version of Response.resume.
 */
/* v8 ignore start - requires actual LLM calls, tested in E2E */
async function _instrumentedResponseResume<F>(
  this: Response<F>,
  content: UserContent,
): Promise<Response<F>> {
  const span = new Span(`Response.resume ${this.modelId}`);
  span.start();

  span.set({
    "mirascope.type": "response_resume",
    "mirascope.response.model_id": this.modelId,
    "mirascope.response.provider_id": this.providerId,
  });

  const otelSpan = span.otelSpan;

  const execute = async (): Promise<Response<F>> => {
    try {
      const result = await _ORIGINAL_RESPONSE_RESUME.call(this, content);
      attachResponseAttributes(span, result);
      return result;
    } catch (error) {
      if (error instanceof Error) {
        span.recordException(error);
      } else {
        span.error(String(error));
      }
      throw error;
    } finally {
      span.finish();
    }
  };

  // Run within the span's context so child spans are properly linked
  return otelSpan
    ? otelContext.with(
        otelTrace.setSpan(otelContext.active(), otelSpan),
        execute,
      )
    : execute();
}
/* v8 ignore end */

/**
 * Replace Response.resume with the instrumented wrapper.
 */
export function wrapResponseResume(): void {
  if (_RESPONSE_RESUME_WRAPPED) {
    return;
  }
  Response.prototype.resume = _instrumentedResponseResume;
  _RESPONSE_RESUME_WRAPPED = true;
}

/**
 * Restore the original Response.resume implementation.
 */
export function unwrapResponseResume(): void {
  if (!_RESPONSE_RESUME_WRAPPED) {
    return;
  }
  Response.prototype.resume = _ORIGINAL_RESPONSE_RESUME;
  _RESPONSE_RESUME_WRAPPED = false;
}

// =============================================================================
// StreamResponse.resume instrumentation
// =============================================================================

/**
 * Instrumented version of StreamResponse.resume.
 */
/* v8 ignore start - requires actual LLM calls, tested in E2E */
async function _instrumentedStreamResponseResume<F>(
  this: StreamResponse<F>,
  content: UserContent,
): Promise<StreamResponse<F>> {
  const span = new Span(`StreamResponse.resume ${this.modelId}`);
  span.start();

  span.set({
    "mirascope.type": "response_resume",
    "mirascope.response.model_id": this.modelId,
    "mirascope.response.provider_id": this.providerId,
  });

  const otelSpan = span.otelSpan;

  const execute = async (): Promise<StreamResponse<F>> => {
    try {
      const result = await _ORIGINAL_STREAM_RESPONSE_RESUME.call(this, content);
      // Note: StreamResponse attributes will be attached after stream is consumed
      return result;
    } catch (error) {
      if (error instanceof Error) {
        span.recordException(error);
      } else {
        span.error(String(error));
      }
      throw error;
    } finally {
      span.finish();
    }
  };

  // Run within the span's context so child spans are properly linked
  return otelSpan
    ? otelContext.with(
        otelTrace.setSpan(otelContext.active(), otelSpan),
        execute,
      )
    : execute();
}
/* v8 ignore end */

/**
 * Replace StreamResponse.resume with the instrumented wrapper.
 */
export function wrapStreamResponseResume(): void {
  if (_STREAM_RESPONSE_RESUME_WRAPPED) {
    return;
  }
  StreamResponse.prototype.resume = _instrumentedStreamResponseResume;
  _STREAM_RESPONSE_RESUME_WRAPPED = true;
}

/**
 * Restore the original StreamResponse.resume implementation.
 */
export function unwrapStreamResponseResume(): void {
  if (!_STREAM_RESPONSE_RESUME_WRAPPED) {
    return;
  }
  StreamResponse.prototype.resume = _ORIGINAL_STREAM_RESPONSE_RESUME;
  _STREAM_RESPONSE_RESUME_WRAPPED = false;
}

// =============================================================================
// ContextResponse.resume instrumentation
// =============================================================================

/**
 * Instrumented version of ContextResponse.resume.
 */
/* v8 ignore start - requires actual LLM calls, tested in E2E */
async function _instrumentedContextResponseResume<DepsT, F>(
  this: ContextResponse<DepsT, F>,
  ctx: Context<DepsT>,
  content: UserContent,
): Promise<ContextResponse<DepsT, F>> {
  const span = new Span(`ContextResponse.resume ${this.modelId}`);
  span.start();

  span.set({
    "mirascope.type": "response_resume",
    "mirascope.response.model_id": this.modelId,
    "mirascope.response.provider_id": this.providerId,
  });

  const otelSpan = span.otelSpan;

  const execute = async (): Promise<ContextResponse<DepsT, F>> => {
    try {
      const result = await _ORIGINAL_CONTEXT_RESPONSE_RESUME.call(
        this,
        ctx,
        content,
      );
      attachResponseAttributes(span, result);
      return result as ContextResponse<DepsT, F>;
    } catch (error) {
      if (error instanceof Error) {
        span.recordException(error);
      } else {
        span.error(String(error));
      }
      throw error;
    } finally {
      span.finish();
    }
  };

  // Run within the span's context so child spans are properly linked
  return otelSpan
    ? otelContext.with(
        otelTrace.setSpan(otelContext.active(), otelSpan),
        execute,
      )
    : execute();
}
/* v8 ignore end */

/**
 * Replace ContextResponse.resume with the instrumented wrapper.
 */
export function wrapContextResponseResume(): void {
  if (_CONTEXT_RESPONSE_RESUME_WRAPPED) {
    return;
  }
  ContextResponse.prototype.resume = _instrumentedContextResponseResume;
  _CONTEXT_RESPONSE_RESUME_WRAPPED = true;
}

/**
 * Restore the original ContextResponse.resume implementation.
 */
export function unwrapContextResponseResume(): void {
  if (!_CONTEXT_RESPONSE_RESUME_WRAPPED) {
    return;
  }
  ContextResponse.prototype.resume = _ORIGINAL_CONTEXT_RESPONSE_RESUME;
  _CONTEXT_RESPONSE_RESUME_WRAPPED = false;
}

// =============================================================================
// ContextStreamResponse.resume instrumentation
// =============================================================================

/**
 * Instrumented version of ContextStreamResponse.resume.
 */
/* v8 ignore start - requires actual LLM calls, tested in E2E */
async function _instrumentedContextStreamResponseResume<DepsT, F>(
  this: ContextStreamResponse<DepsT, F>,
  ctx: Context<DepsT>,
  content: UserContent,
): Promise<ContextStreamResponse<DepsT, F>> {
  const span = new Span(`ContextStreamResponse.resume ${this.modelId}`);
  span.start();

  span.set({
    "mirascope.type": "response_resume",
    "mirascope.response.model_id": this.modelId,
    "mirascope.response.provider_id": this.providerId,
  });

  const otelSpan = span.otelSpan;

  const execute = async (): Promise<ContextStreamResponse<DepsT, F>> => {
    try {
      const result = await _ORIGINAL_CONTEXT_STREAM_RESPONSE_RESUME.call(
        this,
        ctx,
        content,
      );
      // Note: StreamResponse attributes will be attached after stream is consumed
      return result as ContextStreamResponse<DepsT, F>;
    } catch (error) {
      if (error instanceof Error) {
        span.recordException(error);
      } else {
        span.error(String(error));
      }
      throw error;
    } finally {
      span.finish();
    }
  };

  // Run within the span's context so child spans are properly linked
  return otelSpan
    ? otelContext.with(
        otelTrace.setSpan(otelContext.active(), otelSpan),
        execute,
      )
    : execute();
}
/* v8 ignore end */

/**
 * Replace ContextStreamResponse.resume with the instrumented wrapper.
 */
export function wrapContextStreamResponseResume(): void {
  if (_CONTEXT_STREAM_RESPONSE_RESUME_WRAPPED) {
    return;
  }
  ContextStreamResponse.prototype.resume =
    _instrumentedContextStreamResponseResume;
  _CONTEXT_STREAM_RESPONSE_RESUME_WRAPPED = true;
}

/**
 * Restore the original ContextStreamResponse.resume implementation.
 */
export function unwrapContextStreamResponseResume(): void {
  if (!_CONTEXT_STREAM_RESPONSE_RESUME_WRAPPED) {
    return;
  }
  ContextStreamResponse.prototype.resume =
    _ORIGINAL_CONTEXT_STREAM_RESPONSE_RESUME;
  _CONTEXT_STREAM_RESPONSE_RESUME_WRAPPED = false;
}

// =============================================================================
// Convenience functions to wrap/unwrap all methods
// =============================================================================

/**
 * Wrap all Response resume methods with instrumentation.
 */
export function wrapAllResponseResumeMethods(): void {
  wrapResponseResume();
  wrapStreamResponseResume();
  wrapContextResponseResume();
  wrapContextStreamResponseResume();
}

/**
 * Unwrap all Response resume methods, restoring original implementations.
 */
export function unwrapAllResponseResumeMethods(): void {
  unwrapResponseResume();
  unwrapStreamResponseResume();
  unwrapContextResponseResume();
  unwrapContextStreamResponseResume();
}

/**
 * Check if any Response resume method is instrumented.
 */
export function isResponseResumeInstrumented(): boolean {
  return (
    _RESPONSE_RESUME_WRAPPED ||
    _STREAM_RESPONSE_RESUME_WRAPPED ||
    _CONTEXT_RESPONSE_RESUME_WRAPPED ||
    _CONTEXT_STREAM_RESPONSE_RESUME_WRAPPED
  );
}
